build.assets/Makefile

```diff
@@ -107,6 +107,64 @@ build-binaries-fips: buildbox-centos7-fips webassets
 		make -C $(SRCDIR)/e ADDFLAGS='$(ADDFLAGS)' VERSION=$(VERSION) GITTAG=v$(VERSION) PIV=$(PIV) FIPS=yes clean full
 
 #
+# Build the buildbox thirdparty components. This rarely needs to be rebuilt and is
+# slow to build, so it is done separately from the main buildbox
+#
+.PHONY: buildbox-thirdparty
+buildbox-thirdparty:
+	docker buildx build \
+		--cache-from $(BUILDBOX_THIRDPARTY) \
+		--cache-to type=inline \
+		$(if $(PUSH),--push,--load) \
+		--tag $(BUILDBOX_THIRDPARTY) \
+		-f buildbox/Dockerfile-thirdparty \
+		buildbox
+
+#
+# A generic build rule to build a stage of Dockerfile-thirdparty based
+# on the $(STAGE) variable. These stage builds are used for development
+# of the thirdparty buildbox, whether to configure crosstool-NG
+# (see config/buildbox-ng), or when adding additional third party
+# libraries using either the compilers stage or tplibs stage.
+#
+.PHONY: buildbox-thirdparty-stage
+buildbox-thirdparty-stage:
+	docker buildx build \
+		--load \
+		--tag buildbox-thirdparty-$(STAGE):$(BUILDBOX_VERSION) \
+		-f buildbox/Dockerfile-thirdparty \
+		--target $(STAGE) \
+		buildbox
+
+.PHONY: buildbox-thirdparty-crosstoolng
+buildbox-thirdparty-crosstoolng: STAGE=crosstoolng
+buildbox-thirdparty-crosstoolng: buildbox-thirdparty-stage
+
+.PHONY: buildbox-thirdparty-compilers
+buildbox-thirdparty-compilers: STAGE=compilers
+buildbox-thirdparty-compilers: buildbox-thirdparty-stage
+
+.PHONY: buildbox-thirdparty-tplibs
+buildbox-thirdparty-tplibs: STAGE=tplibs
+buildbox-thirdparty-tplibs: buildbox-thirdparty-stage
+
+#
+# Build the buildbox-ng using the pre-built third party components from the
+# buildbox-thirdparty image
+#
+.PHONY: buildbox-ng
+buildbox-ng:
+	docker buildx build \
+		--build-arg THIRDPARTY_IMAGE=$(BUILDBOX_THIRDPARTY) \
```

Comment:
This is tricky as it will always pick-up remote images instead of the local one. I had to modify this line to run it as the remove image doesn't exist.

Reply:
I have not had that problem, but I guess it has something to do with the builder? I think I might be using the original docker driver not the containerd one, but I'm not really sure what I'm seeing with docker once I had to switch to the Mac. I'm using colima, and spent a bit of time stuffing around with the runtime, snapshotter, etc but nothing made sense to me in the end.

I'm really not sure how I'm meant to set this stuff up if it will not look in the local image cache as it does for me but not you. Also in a way that works for devs on Macs and on Linux on CI.

Reply:
> I'm really not sure how I'm meant to set this stuff up

I'd just rename to local build to match the expected name here. In this case when you build locally the local container will be picked up instead of the remote one.

Reply:
I can't make that work. This is a "common" issue with the `docker-container` buildkit driver - it does not have access to the docker image cache except to store built images there when run with `--load`. I have been using the `docker` driver so this has not been an issue, but I ran the following and at the moment I cannot make this work with the `docker-container` driver:

```
$ docker buildx create foo --driver docker-container --use
$ docker images
ghcr.io/gravitational/teleport-buildbox-thirdparty   teleport17         bd51cf80045f   21 hours ago   2.08GB

# ... at this point I have thirdparty image. I give it a "local" name as you suggest:
$ docker tag ghcr.io/gravitational/teleport-buildbox-thirdparty:teleport17 teleport-buildbox-thirdparty:teleport17

# Now I try to build buildbox-ng with this image:
$ make -C build.assets buildbox-ng BUILDBOX_THIRDPARTY=teleport-buildbox-thirdparty:teleport17
...
Dockerfile:26
--------------------
  24 |     # Reference the thirdparty image for copying from later.
  25 |
  26 | >>> FROM ${THIRDPARTY_IMAGE} AS thirdparty
  27 |
  28 |     # ----------------------------------------------------------------------------
--------------------
ERROR: failed to solve: teleport-buildbox-thirdparty:teleport17: failed to resolve source metadata for docker.io/library/teleport-buildbox-thirdparty:teleport17: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

As you can see, it tries to resolve the "local" image name relative to `docker.io/library`.

I can't see any way to make this work using the `docker-container` driver. This is a long-standing issue (I dont want to link it here because I don't want that issue back-linking to this PR: moby buildkit issues/2343#issuecomment-1867944612) with no real resolution right now except to use the `docker` driver instead. But there is no standard name for the builder/context that uses that driver so it is not as simple as just telling docker buildx to use the `docker` driver.

Can you try building this using a build driver that uses the `docker` driver?

Reply:
I experimented with the containerd-snapshotter image cache, which can store multi-arch images, but it still doesn't work - it does not look in the local image cache at all when you are using the `docker-containerd` driver. @jakule If you did manage to get this working somehow, I'd really like to know what you did.